File: tools/gui/src/domain/viewpointCriteria.ts
/**
 * The criteria-tree builder's working model: one recursive tree shape reused for query
 * filtering, neighbor inclusion, style-rule match mode, and matrix axis criteria — mirrors
 * `src/domain/viewpoint_criteria.py` field for field. Each node carries a UI-only `id`
 * (never serialized) so Vue can key/target individual nodes; `viewpointCriteriaSerialization.ts`
 * is the counterpart that converts to/from the canonical wire mapping.
 */

import type { AggregateKind, DerivedAttributeNode, QueryBindingNode, QueryParameterNode } from './viewpointBindings'

let uidCounter = 0
export const nextNodeId = (): string => `n${++uidCounter}`

export type Conjunction = 'and' | 'or'
export type Comparator = 'eq' | 'neq' | 'in' | 'not_in' | 'exists' | 'absent' | 'lt' | 'lte' | 'gt' | 'gte' | 'like' | 'ilike'
export type IncidentDirection = 'outgoing' | 'incoming' | 'either'
export type GroupKind = 'entity' | 'connection'
export type ValueRefKind = 'literal' | 'self' | 'source' | 'target' | 'binding' | 'parameter'
export type Quantifier = 'any' | 'all'

export const NUMERIC_COMPARATORS: readonly Comparator[] = ['lt', 'lte', 'gt', 'gte']
export const STRING_PATTERN_COMPARATORS: readonly Comparator[] = ['like', 'ilike']
export const RESERVED_ENTITY_PATHS: readonly string[] = [
  'id', 'name', 'type', 'specialization', 'group', 'domain', 'subdomain', 'status', 'version',
]
export const RESERVED_CONNECTION_PATHS: readonly string[] = ['id', 'type', 'specialization']

export interface ValueRefLiteral {
  readonly kind: 'literal'
  readonly literal: unknown
}
export interface ValueRefAttribute {
  readonly kind: 'self' | 'source' | 'target'
  readonly attribute: string
}
export interface ValueRefBinding {
  readonly kind: 'binding'
  readonly binding: string
  readonly project: string | null
  readonly aggregate: AggregateKind | null
  readonly quantifier: Quantifier | null
}
export interface ValueRefParameter {
  readonly kind: 'parameter'
  readonly parameter: string
}
export type ValueRef = ValueRefLiteral | ValueRefAttribute | ValueRefBinding | ValueRefParameter

export const literalValue = (literal: unknown): ValueRef => ({ kind: 'literal', literal })
export const selfValue = (attribute: string): ValueRef => ({ kind: 'self', attribute })
export const endpointValue = (endpoint: 'source' | 'target', attribute: string): ValueRef => ({
  kind: endpoint, attribute,
})
export const bindingValue = (
  binding: string,
  options: { project?: string | null; aggregate?: AggregateKind | null; quantifier?: Quantifier | null } = {},
): ValueRef => ({
  kind: 'binding', binding, project: options.project ?? null, aggregate: options.aggregate ?? null,
  quantifier: options.quantifier ?? null,
})
export const parameterValue = (parameter: string): ValueRef => ({ kind: 'parameter', parameter })

export interface ConditionNode {
  readonly kind: 'condition'
  readonly id: string
  attribute: string
  comparator: Comparator
  value: ValueRef
  negate: boolean
}

export interface IncidentNode {
  readonly kind: 'incident'
  readonly id: string
  direction: IncidentDirection
  negate: boolean
  /** direct | derived | both — 'both' is the union of the direct and derived incident
   * sets, computed before any negation. Load-bearing semantics: always serialized. */
  traversal: IncidentTraversal
  connectionCriteria: GroupNode | null
  endpointCriteria: GroupNode | null
}

export interface GroupNode {
  readonly kind: 'group'
  readonly id: string
  readonly groupKind: GroupKind
  conjunction: Conjunction
  negate: boolean
  children: CriteriaNode[]
}

export type CriteriaNode = ConditionNode | IncidentNode | GroupNode

export const mkCondition = (attribute = 'type', comparator: Comparator = 'eq'): ConditionNode => ({
  kind: 'condition', id: nextNodeId(), attribute, comparator, value: literalValue(''), negate: false,
})

export const mkGroup = (groupKind: GroupKind, conjunction: Conjunction = 'and'): GroupNode => ({
  kind: 'group', id: nextNodeId(), groupKind, conjunction, negate: false, children: [],
})

export const mkIncident = (): IncidentNode => ({
  kind: 'incident', id: nextNodeId(), direction: 'either', negate: false, traversal: 'direct',
  connectionCriteria: mkGroup('connection'), endpointCriteria: mkGroup('entity'),
})

export type RelationshipTraversal = 'direct' | 'derived' | 'both'
export type IncidentTraversal = 'direct' | 'derived' | 'both'

export interface NeighborInclusionNode {
  readonly id: string
  direction: IncidentDirection
  connectionCriteria: GroupNode | null
  neighborCriteria: GroupNode | null
  traversal: RelationshipTraversal
  includePotential: boolean
  maxHops: number | null
}

export const mkNeighborInclusion = (): NeighborInclusionNode => ({
  id: nextNodeId(), direction: 'either', connectionCriteria: mkGroup('connection'), neighborCriteria: mkGroup('entity'),
  traversal: 'direct', includePotential: false, maxHops: null,
})

export interface ConnectionSelectionNode {
  enabled: boolean
  criteria: GroupNode
  traversal: RelationshipTraversal
  includePotential: boolean
  maxHops: number | null
}

export const mkConnectionSelection = (): ConnectionSelectionNode => ({
  enabled: true, criteria: mkGroup('connection'), traversal: 'direct', includePotential: false, maxHops: null,
})

export interface ExecutableQueryNode {
  querySchema: number
  entityCriteria: GroupNode
  includeConnected: NeighborInclusionNode[]
  connections: ConnectionSelectionNode
  bindings: QueryBindingNode[]
  parameters: QueryParameterNode[]
  derived: DerivedAttributeNode[]
}

export const QUERY_SCHEMA_VERSION = 1

export const mkQuery = (): ExecutableQueryNode => ({
  querySchema: QUERY_SCHEMA_VERSION,
  entityCriteria: mkGroup('entity'),
  includeConnected: [],
  connections: mkConnectionSelection(),
  bindings: [],
  parameters: [],
  derived: [],
})

File: src/domain/viewpoint_criteria.py
"""The criteria engine: reusable condition trees for viewpoint query filtering, neighbor
inclusion, and style-rule matching.

``EntityCriteriaGroup``/``ConnectionCriteriaGroup`` are the ONE condition-building concept
used everywhere a boolean tree of attribute predicates appears — query filters, style-rule
``mode="match"`` criteria, and matrix axis criteria all reuse these same types rather than
parallel structures. Pure shapes only: parsing lives in ``viewpoint_criteria_parsing.py``,
registry-aware validation in ``viewpoint_criteria_validation.py``, evaluation semantics are
implemented by the evaluator.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

Conjunction = Literal["and", "or"]
Comparator = Literal["eq", "neq", "in", "exists", "absent", "lt", "lte", "gt", "gte"]
ValueRefKind = Literal["literal", "attribute_of_self", "attribute_of_endpoint", "binding", "parameter"]
IncidentDirection = Literal["outgoing", "incoming", "either"]
RelationshipTraversal = Literal["direct", "derived"]

# Exactly today's operator vocabulary — the criteria-tree redesign restructures how
# conditions nest, it does not expand the comparator set. There is deliberately no
# `not_in`: per-condition `negate` already expresses it as `in` + negate, keeping one
# spelling per meaning.
NUMERIC_OPERATORS: frozenset[str] = frozenset({"lt", "lte", "gt", "gte"})
NUMERIC_ATTRIBUTE_TYPES: frozenset[str] = frozenset({"integer", "number", "date"})
VALID_COMPARATORS: frozenset[str] = frozenset({"eq", "neq", "in", "exists", "absent"}) | NUMERIC_OPERATORS
VALID_VALUE_REF_KINDS: frozenset[str] = frozenset(
    {"literal", "attribute_of_self", "attribute_of_endpoint", "binding", "parameter"}
)
VALID_INCIDENT_DIRECTIONS: frozenset[str] = frozenset({"outgoing", "incoming", "either"})
VALID_CONJUNCTIONS: frozenset[str] = frozenset({"and", "or"})

# Addressable properties: reserved read-model paths, resolved before
# the effective schema. None are numeric/date, so numeric comparators are always a save-mode
# error against a reserved path (`version` is explicitly "string comparators only").
RESERVED_ENTITY_PATHS: frozenset[str] = frozenset(
    {"id", "name", "type", "specialization", "group", "domain", "subdomain", "status", "version"}
)
RESERVED_CONNECTION_PATHS: frozenset[str] = frozenset({"id", "type", "specialization"})


@dataclass(frozen=True)
class ValueRef:
    """A condition's comparison value: a literal, or a reference to another attribute.

    ``attribute_of_self`` compares against another attribute on the SAME entity/connection
    being evaluated (e.g. ``end_date >= start_date``). ``attribute_of_endpoint`` is valid
    only within a connection condition, and reads an attribute off the source or target
    entity (e.g. ``strength >= target.threshold``).
    """

    kind: ValueRefKind = "literal"
    literal: object = None
    attribute: str | None = None  # required when kind != "literal"
    endpoint: Literal["source", "target"] | None = None  # required when kind == "attribute_of_endpoint"
    binding: str | None = None
    parameter: str | None = None
    project: str | None = None
    aggregate: Literal["count", "sum", "avg", "min", "max"] | None = None
    quantifier: Literal["any", "all"] | None = None


@dataclass(frozen=True)
class AttributeCondition:
    attribute: str  # dotted path
    comparator: Comparator
    value: ValueRef = ValueRef()
    negate: bool = False  # strict logical complement, including a missing attribute


@dataclass(frozen=True)
class EntityCriteriaGroup:
    conjunction: Conjunction = "and"
    children: "tuple[AttributeCondition | IncidentConnectionCondition | EntityCriteriaGroup, ...]" = ()
    negate: bool = False


@dataclass(frozen=True)
class ConnectionCriteriaGroup:
    conjunction: Conjunction = "and"
    children: "tuple[AttributeCondition | ConnectionCriteriaGroup, ...]" = ()
    negate: bool = False


@dataclass(frozen=True)
class IncidentConnectionCondition:
    """Entity-only predicate: "this entity has (or, negated, does not have) at least one
    incident connection matching ``connection_criteria``/``direction`` whose OTHER endpoint
    matches ``endpoint_criteria``." Fully criteria-based on both legs of the hop; recursive
    via ``endpoint_criteria``, bounded by save-time depth-cap validation.
    """

    connection_criteria: ConnectionCriteriaGroup | None = None  # None = any connection
    direction: IncidentDirection = "either"
    endpoint_criteria: EntityCriteriaGroup | None = None  # None = any entity
    negate: bool = False
    traversal: RelationshipTraversal = "direct"
    include_potential: bool = False
    max_hops: int | None = None


@dataclass(frozen=True)
class ConnectionSelection:
    """Which connections a query displays, within the structural invariant: a
    connection is included only if both its source and target entities are in the included
    entity set. ``criteria`` narrows within that set; it can never widen past it.
    """

    enabled: bool = True
    criteria: ConnectionCriteriaGroup = ConnectionCriteriaGroup()


@dataclass(frozen=True)
class NeighborInclusion:
    """Additive population term: include entities matching ``neighbor_criteria`` that
    are connected — by a connection matching ``connection_criteria``, in ``direction``
    relative to the anchor — to at least one entity of the query's PRIMARY result set.
    Anchors are always the primary set; inclusions never chain off other inclusions' results.
    """

    connection_criteria: ConnectionCriteriaGroup | None = None  # None = any connection
    direction: IncidentDirection = "either"  # relative to the anchor
    neighbor_criteria: EntityCriteriaGroup | None = None  # None = any entity
    traversal: RelationshipTraversal = "direct"
    include_potential: bool = False
    max_hops: int | None = None


EntityCriteriaNode = AttributeCondition | IncidentConnectionCondition | EntityCriteriaGroup
ConnectionCriteriaNode = AttributeCondition | ConnectionCriteriaGroup
